### PR TITLE
[SYCL-MLIR] Simplify polygeist and sycl conversion passes

### PR DESCRIPTION
**Pull request by victor-eds**
This includes:

- Having a single <PassName>.h and <PassName>.cpp file for each pass;
- Declare passes code in the `mlir` namespace, enabling use of auto-generated factory functions;
- Use pass-specific options for configuration instead of `LowerToLLVMOptions` and letting the passes fill those options as needed;
- Drop option controlling "bare pointer conversion" and asserting it is set when calling function populating patterns, as our passes rely on such option being passed, making it "not an option".

### CODE REVIEW

File: mlir-sycl/include/mlir/Conversion/SYCLPasses.td [hunk 1]
```diff
@@ -21,12 +21,7 @@ def ConvertSYCLToLLVM : Pass<"convert-sycl-to-llvm", "ModuleOp"> {
     See docs/SYCLToLLVMDialectConversion/ for more details.
     TODO: add docs referenced above.
   }];
-  let constructor = "mlir::sycl::createConvertSYCLToLLVMPass()";
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Conversion/SYCLPasses.td:24]: Defining the pass in the `mlir` namespace allows using the auto-generated code.

File: mlir-sycl/include/mlir/Conversion/SYCLToGPU/SYCLToGPUPass.h [hunk 2]
```diff
@@ -1,36 +0,0 @@
...
-template <typename T> class OperationPass;
-
-#define GEN_PASS_DECL_CONVERTSYCLTOGPU
-#include "mlir/Conversion/SYCLPasses.h.inc"
-#undef GEN_PASS_DECL_CONVERTSYCLTOGPU
-
-namespace sycl {
-
-/// Creates a pass to convert SYCL operations to the GPU dialect.
-std::unique_ptr<OperationPass<ModuleOp>> createConvertSYCLToGPUPass();
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Conversion/SYCLToGPU/SYCLToGPUPass.h:31]: No need for any of these anymore

File: mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp [hunk 3]
```diff
@@ -248,12 +256,41 @@ void addGridOpPatterns(RewritePatternSet &patterns, MLIRContext *context) {
 }
 } // namespace
 
-void mlir::sycl::populateSYCLToGPUConversionPatterns(
-    RewritePatternSet &patterns) {
+void mlir::populateSYCLToGPUConversionPatterns(RewritePatternSet &patterns) {
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp:259]: Change namespace

File: mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp [hunk 4]
```diff
@@ -248,12 +256,41 @@ void addGridOpPatterns(RewritePatternSet &patterns, MLIRContext *context) {
...
+
+  target
+      .addIllegalOp<SYCLWorkGroupIDOp, SYCLNumWorkItemsOp, SYCLWorkGroupSizeOp,
+                    SYCLLocalIDOp, SYCLGlobalIDOp, SYCLSubGroupIDOp,
+                    SYCLNumSubGroupsOp, SYCLSubGroupSizeOp>();
+
+  if (failed(
+          applyPartialConversion(getOperation(), target, std::move(patterns))))
+    signalPassFailure();
+}
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp:296]: Move here to follow conventions

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 5]
```diff
@@ -1922,8 +1933,10 @@ class GroupGetLocalLinearRangePattern
 // Pattern population
 //===----------------------------------------------------------------------===//
 
-void mlir::sycl::populateSYCLToLLVMTypeConversion(
-    LLVMTypeConverter &typeConverter) {
+void mlir::populateSYCLToLLVMTypeConversion(LLVMTypeConverter &typeConverter) {
+  assert(typeConverter.getOptions().useBarePtrCallConv &&
+         "These patterns only work with bare pointer calling convention");
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:1938]: We assert now

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 6]
```diff
@@ -2003,37 +2016,78 @@ void mlir::sycl::populateSYCLToLLVMTypeConversion(
...
-void mlir::sycl::populateSYCLToLLVMConversionPatterns(
+void mlir::populateSYCLToLLVMConversionPatterns(
     LLVMTypeConverter &typeConverter, RewritePatternSet &patterns) {
+  assert(typeConverter.getOptions().useBarePtrCallConv &&
+         "These pattern only work with bare pointer calling convention");
   populateSYCLToLLVMTypeConversion(typeConverter);
 
   patterns.add<CallPattern>(typeConverter);
   patterns.add<CastPattern>(typeConverter);
-  if (typeConverter.getOptions().useBarePtrCallConv) {
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2012]: Option dropped

File: mlir-sycl/tools/sycl-mlir-opt/sycl-mlir-opt.cpp [hunk 7]
```diff
@@ -37,9 +39,7 @@ int main(int argc, char **argv) {
   // Register passes.
   registerAllPasses();
   sycl::registerSYCLPasses();
-  sycl::registerConvertSYCLToLLVMPass();
-  sycl::registerConvertSYCLToGPUPass();
-  sycl::registerConvertSYCLToSPIRVPass();
+  sycl::registerConversionPasses();
```
Comment (victor-eds) [inline mlir-sycl/tools/sycl-mlir-opt/sycl-mlir-opt.cpp:42]: Nicer

File: polygeist/include/mlir/Conversion/PolygeistToLLVM/PolygeistToLLVM.h [hunk 8]
```diff
@@ -6,8 +6,8 @@
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef MLIR_CONVERSION_POLYGEISTTOLLVM_POLYGEISTTOLLVMPASS_H
-#define MLIR_CONVERSION_POLYGEISTTOLLVM_POLYGEISTTOLLVMPASS_H
+#ifndef MLIR_CONVERSION_POLYGEISTTOLLVM_POLYGEISTTOLLVM_H
+#define MLIR_CONVERSION_POLYGEISTTOLLVM_POLYGEISTTOLLVM_H
```
Comment (victor-eds) [inline polygeist/include/mlir/Conversion/PolygeistToLLVM/PolygeistToLLVM.h:10]: We already had a single file, but with the wrong name

File: polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp [hunk 9]
```diff
@@ -911,23 +912,16 @@ struct GPUReturnOpLowering : public ConvertOpToLLVMPattern<gpu::ReturnOp> {
...
-  ConvertPolygeistToLLVMPass(bool useBarePtrCallConv, bool emitCWrappers,
-                             unsigned indexBitwidth, bool useAlignedAlloc,
-                             const llvm::DataLayout &dataLayout) {
-    this->useBarePtrCallConv = useBarePtrCallConv;
-    this->emitCWrappers = emitCWrappers;
-    this->indexBitwidth = indexBitwidth;
-    this->dataLayout = dataLayout.getStringRepresentation();
-  }
+  using impl::ConvertPolygeistToLLVMBase<
+      ConvertPolygeistToLLVMPass>::ConvertPolygeistToLLVMBase;
```
Comment (victor-eds) [inline polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:916]: Use base class constructors, no need to redefine

File: polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp [hunk 10]
```diff
@@ -911,23 +912,16 @@ struct GPUReturnOpLowering : public ConvertOpToLLVMPattern<gpu::ReturnOp> {
...
 
   void runOnOperation() override {
     ModuleOp m = getOperation();
     const auto &dataLayoutAnalysis = getAnalysis<DataLayoutAnalysis>();
 
     LowerToLLVMOptions options(&getContext(),
                                dataLayoutAnalysis.getAtOrAbove(m));
-    options.useBarePtrCallConv = useBarePtrCallConv;
+    options.useBarePtrCallConv = true;
```
Comment (victor-eds) [inline polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:924]: Always `true`

File: polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp [hunk 11]
```diff
@@ -1040,26 +1031,3 @@ struct ConvertPolygeistToLLVMPass
...
-}
-
-std::unique_ptr<Pass> mlir::createConvertPolygeistToLLVMPass() {
-  // TODO: meaningful arguments to this pass should be specified as
-  // Option<...>'s to the pass in Passes.td. For now, we'll provide some dummy
-  // default values to allow for pass creation.
-  auto dl = llvm::DataLayout("");
-  return std::make_unique<ConvertPolygeistToLLVMPass>(false, false, 64u, false,
-                                                      dl);
-}
```
Comment (victor-eds) [inline polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:1065]: This was messy and is no longer needed (auto generated). Default arguments can be defined as usual in the `.td` file

File: polygeist/tools/cgeist/driver.cc [hunk 12]
```diff
@@ -628,10 +628,9 @@ static int finalize(mlir::MLIRContext &Ctx,
       Module->walk([&](mlir::omp::ParallelOp) { LinkOMP = true; });
       mlir::PassManager PM3(&Ctx);
       PM3.addPass(mlir::sycl::createSYCLMethodToSYCLCallPass());
-      LowerToLLVMOptions Options(&Ctx);
-      Options.dataLayout = DL;
-      Options.useBarePtrCallConv = true;
-      PM3.addPass(createConvertPolygeistToLLVMPass(Options));
+      ConvertPolygeistToLLVMOptions Options;
```
Comment (victor-eds) [inline polygeist/tools/cgeist/driver.cc:631]: Use the pass-specific options and let the pass set the rest.

File: polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp [hunk 13]
```diff
@@ -1040,26 +1031,3 @@ struct ConvertPolygeistToLLVMPass
...
-      useAlignedAlloc, options.dataLayout);
-}
-
-std::unique_ptr<Pass> mlir::createConvertPolygeistToLLVMPass() {
-  // TODO: meaningful arguments to this pass should be specified as
-  // Option<...>'s to the pass in Passes.td. For now, we'll provide some dummy
-  // default values to allow for pass creation.
-  auto dl = llvm::DataLayout("");
-  return std::make_unique<ConvertPolygeistToLLVMPass>(false, false, 64u, false,
-                                                      dl);
```
Comment (victor-eds) [inline polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:1064]: This was indeed wrong, as it was not using the default option for the third argument (should be 0)

**Review by etiotto — APPROVED** [review]
LGTM - Thanks for the cleanup, is good to have consistency with MLIR in how passes are defined!

File: polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp (hunk 10 above)
Comment (victor-eds) [reply polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:924]: We're asserting the type converter has that option set. This is not the option from our pass, but from the generic LLVM conversion options.

File: polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp [hunk 14]
```diff
@@ -938,15 +932,13 @@ struct ConvertPolygeistToLLVMPass
...
-      if (useBarePtrCallConv) {
-        // Keep these at the top; these should be run before the rest of
-        // function conversion patterns.
-        populateReturnOpTypeConversionPattern(patterns, converter);
-        populateCallOpTypeConversionPattern(patterns, converter);
-        populateAnyFunctionOpInterfaceTypeConversionPattern(patterns,
-                                                            converter);
-      }
-      sycl::populateSYCLToLLVMConversionPatterns(converter, patterns);
+      // Keep these at the top; these should be run before the rest of
```
Comment (whitneywhtsang) [inline polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:935]: Not related to this PR. Can we use benefit to ensure these patterns run before rest of the conversion patterns?

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp (hunk 5 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp]: True. Dropped!

File: polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp (hunk 14 above)
Comment (victor-eds) [reply polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:935]: I wish... `populateX` does not receive a `benefit` argument, so we have to rely on ordering here...

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**